# å•ç»†èƒåˆ†æè¸©å‘å®å½•

æ— æƒ…ä½›ç³»æ‰å¿ƒè€é“ 2021-8-29

## åº

**æˆ‘ä»¬ä»ä¸¤ä¸ªé—®é¢˜å…¥æ‰‹ï¼š**

- ç»†å°ç»„ç»‡çš„åˆ’åˆ†é—®é¢˜

  æœ‰æ—¶ï¼Œæˆ‘ä»¬ç ”ç©¶çš„ç›®æ ‡å¯èƒ½æ˜¯ä¸€ä¸ªå…·ä½“ç»†å°çš„å±€éƒ¨ç»„ç»‡ã€‚æ¯”å¦‚ç ”ç©¶è‚ ç»’æ¯›å˜åŒ–ï¼Œç ”ç©¶æ¯›å›Šå‘è‚²è¿‡ç¨‹çš„å˜åŒ–ï¼Œç ”ç©¶å¶ç‰‡æ°”å­”å¼€å…³çš„è°ƒæ§ï¼Œç ”ç©¶é›„æ€§ä¸è‚²è¿‡ç¨‹çš„èŠ±ç²‰è´¥è‚²ã€‚

  ç”±äºæŠ€æœ¯çš„é™åˆ¶ï¼Œæˆ‘ä»¬å¾ˆéš¾ç²¾ç¡®åœ°é€‰æ‹©ç›®æ ‡éƒ¨ä½ï¼Œåªèƒ½æ¨¡ç³Šåœ°å–å°è‚ çš„ä¸€éƒ¨åˆ†ï¼ŒæŒ‘å–ä¸€å¤§ç‰‡çš®è‚¤ï¼Œä½¿ç”¨æ•´ä¸ªå¶ç‰‡ï¼Œä½¿ç”¨æ•´ä¸ªé›„è•Šç­‰ï¼Œç„¶åè¿›è¡Œè½¬å½•ç»„æµ‹åºã€‚åœ¨è¿™æ ·çš„æƒ…å†µä¸‹ï¼Œç›®æ ‡ç»„ç»‡åœ¨æ•´ä¸ªæµ‹åºæ ·æœ¬ä¸­åªå å¾ˆå°çš„ä¸€éƒ¨åˆ†ï¼Œæµ‹åºæ ·æœ¬å¤§æ¦‚ç‡æ— æ³•å‘ˆç°æˆ‘ä»¬å…³å¿ƒçš„ä¿¡æ¯ã€‚

  è¿™å¥½æ¯”æŠŠä¸€å°å‹ºæ©™æ±åŠ åˆ°äº†ä¸€æ¯è‘¡è„é…’é‡Œã€‚ä½ çš„ç›®æ ‡æœ¬æƒ³ç ”ç©¶ä¸åŒæ©™æ±çš„ç”œåº¦æ˜¯å¦æœ‰å·®å¼‚ã€‚ä½†æ©™æ±çš„å‘³é“æ—©è¢«ç¨€é‡Šäº†ï¼Œåè€Œæœ€åæ¯å­é‡Œé¥®æ–™çš„å‘³é“æ›´å¤šåæ˜ çš„æ˜¯è‘¡è„é…’çš„å‘³é“ã€‚

  å¯¹äºè¿™ç§æƒ…å†µï¼Œå¤§éƒ¨åˆ†äººä¹Ÿèƒ½æ¸…æ¥šå·²æœ‰æ–¹æ³•çš„å±€é™ã€‚ä½†é™¤éæœ‰æŠ€æœ¯å¯ä»¥å¯¹ç›®æ ‡ç»„ç»‡è¿›è¡Œç²¾ç¡®å–æ ·ï¼Œæˆ–è€…å¯¹ä»¥æ ·æœ¬ä¸­çš„ç»†èƒä¸ºå•ä½è¿›è¡Œç ”ç©¶ï¼Œå¦åˆ™ä»¥ä¸Šçš„é—®é¢˜æ— æ³•è§£å†³ã€‚

- ç»†èƒçš„å¼‚è´¨æ€§é—®é¢˜

  å“ªæ€•æ˜¯ä¸€ä¸ªå•ä¸€çš„ç»„ç»‡ï¼Œä¹Ÿä¸ä»…ä»…åªæœ‰ä¸€ç§ç»†èƒï¼Œè€Œå®é™…ä¸Šæ˜¯å„ç§ç»†èƒçš„å¤§æ‚çƒ©ã€‚

  ä»¥å°é¼ å¿ƒè„ç»„ç»‡ä¸ºä¾‹ï¼Œå¯èƒ½ç›´è§‰ä¸Šè®¤ä¸ºå¿ƒè„ä¸»è¦æ˜¯ä¸€å›¢ä¼šè·³åŠ¨çš„è‚Œè‚‰ï¼Œä¸»è¦æ˜¯å¿ƒè‚Œç»†èƒï¼ˆCardiomyocytesï¼‰ã€‚ä½†å¦‚æœè¿›è¡Œå•ç»†èƒæ£€æµ‹ï¼ˆå¦‚ä¸‹å›¾ï¼‰ï¼Œå¯ä»¥çœ‹å‡ºå¿ƒè„çš„ç»†èƒæ„æˆå®é™…ä¸Šéå¸¸ä¸°å¯Œï¼Œä¸ä»…æœ‰å¿ƒè‚Œç»†èƒï¼Œè¿˜åŒ…æ‹¬æˆçº¤ç»´ç»†èƒï¼ˆFibroblastsï¼‰ã€å†…çš®ç»†èƒï¼ˆEndothelial cellsï¼‰ã€è¡€ç®¡å†…çš®ç»†èƒï¼ˆVessel Endothelialï¼‰ã€å¿ƒå¤–è†œç»†èƒï¼ˆEpicardialï¼‰ã€å·¨å™¬ç»†èƒï¼ˆMacrophagesï¼‰ã€æ·‹å·´ç»†èƒï¼ˆLymphocytesï¼‰ã€å¹³æ»‘è‚Œç»†èƒï¼ˆSmooth muscleï¼‰ç­‰ï¼Œè€Œæˆçº¤ç»´ç»†èƒåˆ™åˆå¯ä»¥è¿›ä¸€æ­¥åˆ†ä¸ºè‹¥å¹²ä¸ªå­äºšç¾¤ã€‚

  ![img](http://5b0988e595225.cdn.sohucs.com/images/20190605/a8488c2dd3cd422eb398b6ee62fb2e8d.jpeg)

  é—®é¢˜å¾€å¾€å°±å‡ºåœ¨è¿™é‡Œï¼šå¸¸è§„è½¬å½•ç»„æ£€æµ‹çš„æ˜¯æ ·æœ¬ä¸­æ‰€æœ‰ç»†èƒçš„å‡å€¼ï¼Œè€Œæˆ‘ä»¬ç ”ç©¶ç›®æ ‡ï¼ˆä¾‹å¦‚ï¼ŒæŸç§ç–¾ç—…çš„å‘ç”Ÿæœºåˆ¶ï¼‰å´å¯èƒ½åªä¸å…¶ä¸­çš„æŸç§ç»†èƒç›¸å…³ã€‚ä¾‹å¦‚ï¼Œå¦‚æœç ”ç©¶å¿ƒè„çº¤ç»´å˜æ€§ï¼Œæœ€ä½³æ–¹æ¡ˆåº”è¯¥åªé’ˆå¯¹å¿ƒè„æˆçº¤ç»´ç»†èƒè¿›è¡Œæ£€æµ‹å’Œæ¯”è¾ƒï¼Œè€Œæ™®é€šè½¬å½•ç»„å¾€å¾€æ˜¯æ£€æµ‹æ•´ä¸ªå¿ƒè„ç»„ç»‡ã€‚å¦‚æœæˆ‘ä»¬ç ”ç©¶èƒ°è…ºçš„èƒ°å²›ç´ åˆ†æ³Œå˜åŒ–ï¼Œé‚£ä¹ˆç†è®ºä¸Šåº”è¯¥é’ˆå¯¹èƒ°å²›Î±ç»†èƒè¿›è¡Œæ£€æµ‹æ¯”è¾ƒï¼Œè€Œæ™®é€šè½¬å½•ç»„å¯èƒ½æ˜¯å¯¹æ•´ä¸ªèƒ°è…ºè¿›è¡Œæ£€æµ‹ã€‚

  è¿™ç§æœŸæœ›ï¼ˆç ”ç©¶èŒƒå›´åº”è¯¥é™å®šåœ¨ç‰¹å®šç±»å‹ç»†èƒä¸­ï¼‰ä¸ç°å®ï¼ˆåªèƒ½æ£€æµ‹å„ç§æ··åˆç»†èƒçš„å‡å€¼ï¼‰å·¨å¤§è£‚éš™ï¼Œå¯¼è‡´äº†ä»¥å¾€çš„å‘è‚²ã€ç–¾ç—…ã€ç¯å¢ƒåº”ç­”ç­‰çš„ç»„å­¦ç ”ç©¶åªèƒ½å¾—åˆ°ä¸€ä¸ªæ¦‚è¦æ€§çš„ç»“è®ºï¼Œè€Œéš¾ä»¥åœ¨æ›´åŠ ç²¾ç»†åœ°åœ¨ç»†èƒå°ºåº¦ç»™å‡ºä¸€ä¸ªæ¸…æ™°çš„è§£ç­”ã€‚è¦å¼¥è¡¥è¿™ä¸ªè£‚éš™ï¼Œåªèƒ½ä½¿ç”¨å•ç»†èƒæ£€æµ‹çš„æ–¹æ³•ã€‚https://www.sohu.com/a/318595293_278730

æ–°çš„é—®é¢˜éœ€è¦æ–°çš„æ•°æ®ç±»å‹ï¼ˆæ–°çš„æ•°æ®é‡‡é›†æ–¹å¼ï¼‰ï¼Œè€Œæ–°çš„æ•°æ®ç±»å‹éœ€è¦æ ¹æ®ä»–çš„ç‰¹ç‚¹å¼€å‘æ–°çš„æ•°æ®åˆ†ææ–¹å¼ã€‚

2009å¹´åŒ—äº¬å¤§å­¦æ±¤å¯Œé…¬æ•™æˆåœ¨åšå£«åæœŸé—´å‘è¡¨äº†ä¸–ç•Œç¬¬ä¸€ç¯‡å•ç»†èƒmRNAæµ‹åºçš„æ–‡ç« ï¼Œå¼€å¯äº†å•ç»†èƒç»„å­¦æ—¶ä»£ï¼ˆscRNA-seqï¼‰ã€‚https://pubmed.ncbi.nlm.nih.gov/19349980/

**ç›®å‰ä¸»è¦æœ‰ä¸¤ç§ç­–ç•¥æ¥å®ç°å•ç»†èƒæµ‹åºã€‚**

ç¬¬ä¸€ç§ï¼Œä¹Ÿå°±æ˜¯ç›®å‰å¤§å¤šæ•°äººæ‰€æƒ³è±¡çš„é‚£æ ·ï¼Œå°†å•ä¸ªç»†èƒåˆ†ç¦»å‡ºæ¥ï¼Œå¹¶ç‹¬ç«‹æ„å»ºæµ‹åºæ–‡åº“ï¼Œæœ€ç»ˆè¿›è¡Œæµ‹åºçš„è·¯çº¿ã€‚æˆ‘ä»¬å¯ä»¥é€šè¿‡æµå¼ç»†èƒæœ¯ï¼ˆå«å¾®æµä½“èŠ¯ç‰‡ï¼‰ï¼Œæˆ–è€…æ¿€å…‰æ•è·æ˜¾å¾®åˆ‡å‰²ï¼ˆLCMï¼‰æ¥å®ç°ã€‚æµå¼ç»†èƒæœ¯ä¼°è®¡å¤§å®¶æ¯”è¾ƒç†Ÿæ‚‰ï¼Œå°±ä¸å¤šè®²äº†ï¼Œå®ƒä¸»è¦è¿ç”¨äºç»†èƒæ ·å“ã€‚å¯¹äºç»„ç»‡åˆ‡ç‰‡æ ·å“æ¥è¯´ï¼Œä¸»è¦æ˜¯é€šè¿‡LCMæ¥è·å–å•ç»†èƒï¼ŒåŸç†å¯ä»¥è§ä¸‹é¢çš„ç¤ºæ„å›¾ã€‚

è¿‘å¹´æ¥å¤šé‡‡å–ç¬¬äºŒç§ç­–ç•¥ï¼šåŸºäºæ ‡ç­¾ï¼ˆbarcodeï¼‰çš„å•ç»†èƒè¯†åˆ«ã€‚å®ƒçš„ä¸»è¦æ€æƒ³æ˜¯ï¼Œç»™æ¯ä¸ªç»†èƒåŠ ä¸Šç‹¬ä¸€æ— äºŒçš„DNAåºåˆ—ï¼Œè¿™æ ·åœ¨æµ‹åºçš„æ—¶å€™ï¼Œå°±æŠŠæºå¸¦ç›¸åŒbarcodeçš„åºåˆ—è§†ä¸ºæ¥è‡ªåŒä¸€ä¸ªç»†èƒäº†ã€‚è¿™ç§ç­–ç•¥ï¼Œå¯ä»¥é€šè¿‡ä¸€æ¬¡å»ºåº“ï¼Œæµ‹å¾—æ•°ç™¾ä¸Šåƒä¸ªå•ç»†èƒçš„ä¿¡æ¯ã€‚æˆ‘é‡åˆ°çš„æ•°æ®å°±æ˜¯è¿™ä¹ˆæµ‹çš„ã€‚

**é‚£å•ç»†èƒæµ‹åºæ•°æ®æœ‰ä»€ä¹ˆç‰¹ç‚¹å‘¢ï¼Ÿ**

å¤§æ•°æ®çš„å››å¤§ç‰¹ç‚¹å‡ ä¹éƒ½åœ¨å•ç»†èƒæ•°æ®ä¸­ä½“ç°äº†ï¼š

1. æµ·é‡æ€§ 

   å•ç»†èƒçš„æµ·é‡æ€§ä¸ä»…ä½“ç°åœ¨å•æ¬¡æ•è·çš„ç»†èƒæ•°å’ŒåŸºå› æ•°ä¸Šï¼Œè€Œæ˜¯æ¯å¤©ä¸åŒçš„å®éªŒå®¤äº§ç”Ÿå¯¹å•ç»†èƒä¸åŒçš„åˆ»ç”»ï¼Œä¸åŒå™¨å®˜ï¼Œä¸åŒç‰©ç§ï¼Œä¸åŒæŠ€æœ¯å±‚é¢ã€‚

2. å¤šæ ·æ€§ 

   å¤šæ ·æ€§ç¿»è¯‘åˆ°å•ç»†èƒè¿™å°±æ˜¯ä¸åŒçš„æ¨¡æ€ï¼Œä¸ä»…æœ‰è¡¨å¾ä¸°åº¦çš„çŸ©é˜µæ•°æ®ï¼Œè¿˜æœ‰ç©ºé—´æ•°æ®ï¼ˆç©ºé—´è½¬å½•ç»„ç­‰ï¼‰ï¼Œç»“æ„æ•°æ®ã€‚

3. é«˜é€Ÿæ€§å’Œæ˜“å˜æ€§

   é«˜é€Ÿå’Œæ˜“å˜å°±ä¸å¤šè¯´äº†ï¼Œç›®å‰æœ‰ä¸å°‘æ–‡ç« å‘å‡ºæ¥å°±æ˜¯ä¸€ä¸ªæ•°æ®åº“ï¼Œå¾€å¾€æœ‰ä¸€ä¸ªæ–°çš„æŠ€æœ¯å¤´å·ç©å®¶å®Œäº†ï¼ŒäºŒå·åŸºæœ¬æ²¡æœºä¼šäº†ã€‚

https://cloud.tencent.com/developer/article/1646016#

**é‚£å•ç»†èƒæ•°æ®åˆ†æéœ€è¦æˆ‘ä»¬åšä»€ä¹ˆå‘¢ï¼Ÿ**

è¯¦è§æ­¤æ–‡https://pubmed.ncbi.nlm.nih.gov/31217225/ï¼Œä»¥ä¸‹ä¸¤æµç¨‹å›¾æ–¹æ¡†ä»£è¡¨æ•°æ®ï¼Œåœ†æ¡†ä»£è¡¨æ•°æ®å¤„ç†æ–¹æ³•ã€‚å…·ä½“å†…å®¹è¯¦è§æ­£æ–‡ï¼Œéƒ½æœ‰ç°æˆçš„æ–¹æ³•ğŸ¤ª

- Pre-processing

```mermaid
graph TD

A[Single Cell Sequencing Raw Data] -->B(Raw Data Processing) 
B --> H[Count Matrices]

H --> G(Quality Control)
H --> I(Normalization)
H --> J(Data Correction)
H --> E(Feature Selection)
G --> K[Better Data]
I --> K[Better Data]
J --> K[Better Data]
E --> K[Better Data]
```

- Downstream Analysisï¼ˆä¸å…¨ï¼‰


```mermaid
graph TD

A[Better Data] -->B(Visualization) 
B --> C(Clustering)
C --> E(Difference Expression Analysis)
C --> G(Enrichment Analysis)
C --> D(Marker Identification)
C --> F(Pseudotime Trajectory Analysis)
```

æ€»è€Œè¨€ä¹‹ï¼Œå•ç»†èƒæµ‹åºæŠ€æœ¯æ˜¯æˆ‘ä»¬ä¸­å›½äººç ”å‘çš„æ–¹æ³•ï¼Œå•ç»†èƒæ•°æ®åˆ†æä¸å­¦ä¸æ˜¯ä¸­å›½äººï¼ğŸ¶

è¯ä¸å¤šè¯´ï¼Œå¼€å¯æ„‰zhaå¿«xinçš„å•ç»†èƒæ•°æ®åˆ†æï¼ˆåšèœï¼‰è¸©å‘ä¹‹æ—…ã€‚

## Part I ç£¨åˆ€ä¸è¯¯ç æŸ´å·¥â€”â€”é¢„å¤‡çŸ¥è¯†

- Linux

  http://www.ehbio.com/tutorial/%E7%94%9F%E4%BF%A1%E5%AE%9D%E5%85%B8-Linux%E6%95%99%E7%A8%8B.pdf

- Python

  http://www.ehbio.com/tutorial/%E7%94%9F%E4%BF%A1%E5%AE%9D%E5%85%B8Py3_course.pdf

- R

  http://www.ehbio.com/tutorial/%E7%94%9F%E4%BF%A1%E5%AE%9D%E5%85%B8-R%E5%AD%A6%E4%B9%A0%E6%95%99%E7%A8%8B.pdf

## Part II å‡†å¤‡å®¶ä¼™äº‹å„¿â€”â€”æ­å»ºç¯å¢ƒ

### 2.1 æœåŠ¡å™¨å®‰è£…anacondaå¹¶åˆ›å»ºè™šæ‹Ÿç¯å¢ƒ

https://blog.csdn.net/wyf2017/article/details/118676765

### 2.2 åœ¨è™šæ‹Ÿç¯å¢ƒä¸­å®‰è£…R

https://www.jianshu.com/p/2621abd6493a

ä¹‹åå†æ·»åŠ ä¸€äº›channel

https://blog.csdn.net/weixin_39278265/article/details/84782550

### 2.3 å®‰è£…RåŒ…Seuratï¼ˆç¬¬ä¸€ä¸ªå‘ï¼‰

https://satijalab.org/seurat/articles/install.html

æœŸé—´å¯èƒ½ä¼šæŠ¥å„ç§å„æ ·çš„é”™è¯¯ï¼Œéƒ¨åˆ†è§£å†³æ–¹æ³•å¦‚ä¸‹ï¼š

- R ç‰ˆæœ¬é—®é¢˜ï¼šå‚è€ƒ2.2ï¼ŒæŠŠéœ€è¦çš„r-baseç‰ˆæœ¬ï¼Œä¹Ÿå°±æ˜¯Rå†ä¸‹ä¸€éã€‚

- éƒ¨åˆ†åŒ…ä¸ä¸€è‡´å¯¼è‡´å®ƒä»¬å®‰ä¸ä¸Šé—®é¢˜ï¼šå‚è€ƒ2.2ï¼Œç”¨condaï¼Œbioconductorè¯•è¯•ï¼Œä¹Ÿå¯ä»¥ç”¨æºæ–‡ä»¶æœ¬åœ°å®‰è£…æ–¹å¼ï¼š

  R: install.packages(â€œxxxx.xxxâ€,repo=NULL,type=â€sourceâ€)

  conda: conda install --use-local xxx.xxx

- éƒ¨åˆ†åŒ…æŠ½é£æŠ¥é”™ï¼ˆRæ— æ³•è¯»å…¥å®ƒçš„libç­‰é—®é¢˜ï¼‰ï¼šå¸äº†é‡è£…ä¸€éå°±å¥½ï¼Œå°±æ˜¯æ¬ æå•Šï¼

### 2.4 å®‰è£…jupyter labå¹¶æ·»åŠ R kernel

conda install jupyter

è¿›å…¥R

install.packages('devtools')

devtools::install_github('IRkernel/IRkernel')

IRkernel::installspec(name = 'irç‰ˆæœ¬', displayname = 'R+ç‰ˆæœ¬')

è¿™æ­¥ç½‘ä¸Šå¤šæ•°çš„æ•™å­¦æ˜¯ä¸å¥½çš„

### 2.5 è¿œç¨‹ç™»å…¥jupyter lab

https://zhuanlan.zhihu.com/p/367020783

å½“ç„¶ä¹Ÿå¯ä»¥åŠ ä¸Šnohup

### 2.6 å®‰è£…Cell Rangerå¹¶ä¸‹è½½å‚è€ƒåŸºå› ç»„

https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/installation

### 2.7 æ ¹æ®éœ€è¦å®‰è£…å…¶ä»–åŒ…

scanpy monocle3 ç­‰

## Part III æ´—èœâ€”â€”raw dataå¤„ç†

https://www.jianshu.com/p/51e0ef0c7301

æ³¨æ„è¿™é‡Œä¸éœ€è¦catï¼Œè¿™æ ·å°±å¯ä»¥  --sample=CC5F0-1,CC5F0-2,CC5F0-3,CC5F0-4

å¦‚æœæœ‰å¥½å‡ ä¸ªbatchï¼Œå¯ä»¥å†™ä¸€ä¸ªforå¾ªç¯æ”¾åœ¨.shæ–‡ä»¶ã€‚

è¿™ä¸ªæ•°æ®é‡ä¸€èˆ¬å¾ˆå¤§ï¼Œé€šå¸¸è¦åŠ nohupè·‘ï¼Œè¿è¡Œæ—¶é—´ä»¥å°æ—¶æˆ–å¤©æ¥è®¡ç®—ã€‚



## Part IV æ‹©èœâ€”â€”è¡¨è¾¾æ•°æ®é¢„å¤„ç†

ä»¥ä¸‹å†…å®¹åŸºäºä¸¤ç¯‡å®˜æ–¹å‚è€ƒæ•™å­¦ï¼š

https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html

https://satijalab.org/seurat/articles/pbmc3k_tutorial.html

è¿™é‡Œç”¨pythonè·‘scanpyï¼Œå› ä¸ºä¹‹åè¦ç”¨åˆ°ä¸€ä¸ªç‰¹æ®Šçš„pythonåŒ…ã€‚

### 4.1 è½½å…¥å¿…é¡»çš„åŒ…

```python
import numpy as np
import pandas as pd
import scanpy as sc
import matplotlib.pyplot as plt
import os
from anndata import AnnData
from scipy.sparse import issparse
import desc
from matplotlib.pyplot import rc_context

os.chdir('/s/e/tanbowen')#åˆ‡æ¢å·¥ä½œè·¯å¾„
```

### 4.2 è¯»å–è¡¨è¾¾çŸ©é˜µã€è®¡ç®—çº¿ç²’ä½“åŸºå› ç›¸å…³ï¼ˆä¸‹æ­¥åšè´¨é‡æ§åˆ¶ï¼‰ã€ç”»violinå›¾ã€åšè´¨é‡æ§åˆ¶ï¼ˆæ»¤ç»†èƒï¼‰

```python
Name = []
a = ()
for i in range(1,5):
    name = 'Donor'+str(i)
    path = '/s/e/tanbowen/cellranger_'+name+'/outs/filtered_feature_bc_matrix/'
    adata = sc.read_10x_mtx(
        path,  # the directory with the `.mtx` file
        var_names='gene_symbols',                # use gene symbols for the variable names (variables-axis index)
        cache=True) 
    adata.var['mt'] = adata.var_names.str.startswith('MT-')  # annotate the group of mitochondrial genes as 'mt'
    sc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)
    sc.pp.filter_cells(adata, min_genes=200)
    sc.pp.filter_cells(adata, max_genes=6000)
    adata = adata[adata.obs.pct_counts_mt < 20, :]
    exec("adata_Donor%s = adata"%i)
    Name.append(name)
    a = a+(adata,)
```

**ä¸ºä»€ä¹ˆè¦åšè´¨æ§**ï¼Ÿ

åœ¨ç»†èƒåˆ†ç¦»è¿‡ç¨‹ä¸­çš„ç»†èƒæŸä¼¤æˆ–è€…æ–‡åº“åˆ¶å¤‡çš„å¤±è´¥ï¼ˆæ— æ•ˆçš„é€†è½¬å½•æˆ–è€…PCRæ‰©å¢å¤±è´¥ï¼‰ï¼Œå¾€å¾€ä¼šå¼•å…¥ä¸€äº›ä½è´¨é‡çš„æ•°æ®ã€‚è¿™äº›ä½è´¨é‡çš„æ•°æ®çš„ä¸»è¦ç‰¹ç‚¹æ˜¯ï¼š

1. ç»†èƒæ•´ä½“ä¸Šçš„countså€¼å°‘ï¼ˆåˆ—ï¼‰

2. åŸºå› çš„ä½è¡¨è¾¾ï¼ˆè¡Œï¼‰

3. çº¿ç²’ä½“åŸºå› æˆ–è€…spike-inçš„æ¯”ä¾‹ç›¸å¯¹è¾ƒé«˜

å¦‚æœè¿™äº›æŸä¼¤çš„è¡Œæˆ–è€…åˆ—ï¼Œæ²¡æœ‰è¢«ç§»é™¤çš„è¯ï¼Œå¯èƒ½ä¼šå¯¹ä¸‹æ¸¸çš„åˆ†æç»“æœäº§ç”Ÿå½±å“ã€‚æ‰€ä»¥æˆ‘ä»¬åœ¨è¿›è¡Œåˆ†æä¹‹å‰ï¼Œä¸€å®šè¦ç‡å…ˆç§»é™¤è¿™äº›ä½è´¨é‡çš„è¡Œä¸åˆ—ã€‚ï¼ˆä¸€å¼€å§‹çš„ç†è§£ï¼Œåé¢æ•´ä¸ªæµç¨‹åšå®Œä¹‹åï¼Œæˆ–è®¸ç†è§£ä¼šæ›´å¤šï¼Œé‚£ä¹ˆæ¥ä¸‹æ¥åœ¨åšè¯¦ç»†çš„è¡¥å……ï¼‰

**è´¨æ§çš„æŒ‡æ ‡**

1. æ¯ä¸€ä¸ªç»†èƒæ‰€æœ‰åŸºå› çš„countså€¼ä¹‹å’Œ

   åœ¨æ–‡åº“åˆ¶å¤‡çš„è¿‡ç¨‹ä¸­ï¼Œå¯èƒ½å› ä¸ºç»†èƒçš„è£‚è§£æˆ–cDNAæ•è·å’Œæ‰©æ‹›æ•ˆç‡çš„ä½ä¸‹ï¼Œè€Œä½¿å¾—RNAçš„ä¸¢å¤±ã€‚å…·æœ‰è¾ƒå°çš„countså€¼ä¹‹å’Œçš„ç»†èƒè¢«è®¤ä¸ºæ˜¯ä½è´¨é‡çš„ç»†èƒï¼Œè€ƒè™‘è¢«å»é™¤ã€‚

2. æ¯ä¸€ä¸ªç»†èƒä¸­å•ä¸ªåŸºå› çš„è¡¨è¾¾æ•°é‡

   å¤šæ ·åŒ–çš„è½¬å½•æœ¬å¦‚æœæ²¡æœ‰è¢«æˆåŠŸçš„æ•è·åˆ°ï¼Œå› æ­¤ä»»ä½•ä¸€ä¸ªç»†èƒä¸­æœ‰å¾ˆå°‘çš„åŸºå› è¡¨è¾¾ï¼Œè¢«è®¤ä¸ºæ˜¯ä½è´¨é‡çš„ï¼Œè€ƒè™‘è¢«å»é™¤ã€‚

3. æ¯ä¸€ä¸ªç»†èƒä¸­ï¼Œspike-inåºåˆ—/çº¿ç²’ä½“åŸºå› å æ€»çš„countså€¼çš„æ¯”ä¾‹

   æ¯ä¸ªç»†èƒä¸­æ·»åŠ çš„spike-inåºåˆ—ï¼ˆäººä¸ºæ·»åŠ çš„è¡¨è¾¾é‡çš„å‚ç…§ç³»ï¼‰çš„æµ“åº¦éƒ½æ˜¯ç­‰é‡çš„ã€‚å¦‚æœspike-inçš„æ¯”å€¼å¾ˆé«˜ï¼Œé‚£ä¹ˆå°±æ„å‘³ç€åœ¨å®éªŒçš„è¿‡ç¨‹ä¸­ï¼Œå¤§é‡çš„è½¬å½•æœ¬ä¸¢å¤±ã€‚

   åŒæ ·ï¼Œçº¿ç²’ä½“åŸºå› çš„é«˜æ¯”ä¾‹ï¼Œä¹Ÿæ„å‘³ç€è¿™å¯èƒ½æ˜¯ç”±äºç©¿å­”ç»†èƒçš„ç»†èƒè´¨RNAä¸¢å¤±ï¼Œä»è€Œäº§ç”Ÿä½è´¨é‡çš„ç»†èƒã€‚ç†ç”±æ˜¯ï¼Œåœ¨å­˜åœ¨é€‚åº¦ç»†èƒæŸä¼¤çš„æƒ…å†µä¸‹ï¼Œç»†èƒè†œä¸Šçš„å­”å…è®¸å•ä¸ªè½¬å½•ç‰©åˆ†å­å¤–æ’ï¼ˆä¸¢å¤±ï¼‰ï¼Œä½†è¿‡å°è€Œæ— æ³•ä½¿çº¿ç²’ä½“é€¸å‡ºï¼Œä»è€Œå¯¼è‡´çº¿ç²’ä½“è½¬å½•ç‰©çš„ç›¸å¯¹å¯Œé›†ã€‚
   https://blog.csdn.net/weixin_40640700/article/details/114538295

### 4.3 æ‰¹æ¬¡æ•´åˆã€æ»¤åŸºå› ã€ä¿å­˜æ•°æ®ã€æ ‡å‡†åŒ–ã€æŒ‘é€‰å‡ºé«˜è¡¨è¾¾åŸºå› 

```python
adata_all = adata_Donor1.concatenate(t,batch_key = 'obs', batch_categories = Name)
adata_all.obs.rename(columns={'obs':'batch'},inplace=True)
sc.pp.filter_genes(adata_all, min_cells=3)

adata_all.raw=adata_all

desc.normalize_per_cell(adata_all, counts_per_cell_after=1e4)
desc.log1p(adata_all)
sc.pp.highly_variable_genes(adata_all, n_top_genes = 2000, batch_key = None)
adata_all = adata_all[:, adata_all.var['highly_variable']]
```

## Part V ç‰¹æ®Šå·¥åºâ€”â€”descå»æ‰¹æ¬¡æ•ˆåº”å¹¶èšç±»å¯è§†åŒ–

æ–‡ç« ï¼šhttps://www.nature.com/articles/s41467-020-15851-3

å®˜æ–¹ç½‘å€ï¼šhttps://eleozzr.github.io/desc/

### 5.1 scale_bygroup

æ­£å¸¸åº”è¯¥æ˜¯è¿™æ ·ä¸€æ­¥å°±å¯ä»¥äº†

```python
scale_bygroup(adata_all, groupby='batch', max_value=6)
```

å¯èƒ½ä¼šæŠ¥å…³äºnumpyé”™ã€‚å¯ä»¥å°†å‚ç…§æºç ç¨ä½œä¿®æ”¹

```python
def _log1p(x):
    if issparse(x):
        np.log1p(x.data, out=x.data)
    else:
        np.log1p(x, out=x)
    return x


def log1p(data, copy=False):
    if copy:
        data = data.copy()
    if isinstance(data, AnnData):
        _log1p(data.X)
    else:
        _log1p(data)
    return data if copy else None

def get_mean_var(X):
# - using sklearn.StandardScaler throws an error related to
#   int to long trafo for very large matrices
# - using X.multiply is slower
    if True:
        mean = X.mean(axis=0)
        if issparse(X):
            mean_sq = X.multiply(X).mean(axis=0)
            mean = mean.A1
            mean_sq = mean_sq.A1
        else:
            mean_sq = np.multiply(X, X).mean(axis=0)
        # enforece R convention (unbiased estimator) for variance
        var = (mean_sq - mean**2) * (X.shape[0]/(X.shape[0]-1))
    else:
        from sklearn.preprocessing import StandardScaler
        scaler = StandardScaler(with_mean=False).partial_fit(X)
        mean = scaler.mean_
        # enforce R convention (unbiased estimator)
        var = scaler.var_ * (X.shape[0]/(X.shape[0]-1))
    return mean, var


def scale_bygroup(adata,groupby,max_value=6):
    res=None
    assert isinstance(adata,AnnData),'adata must be AnnData class'
    adata.X=adata.X.toarray() if issparse(adata.X) else adata.X
    if groupby in adata.obs.keys():
        df=pd.Series(adata.obs[groupby],dtype="category")
        for category in df.cat.categories:
            tmp=adata[df==category]
            tmp=tmp.X.copy()
#             tmp=np.asarray(tmp)
            mean0,var0=get_mean_var(tmp)
            sd0=np.sqrt(var0)
            sd0[sd0<=1e-5]=1e-5
            tmp-=mean0
            tmp/=sd0
            if max_value is not None:
                tmp[tmp>max_value]=max_value
            adata.X[df==category]=tmp.copy()
    else:
        print("Warning: The groupby:"+str(groupby)+ "you provided is not exists, we scale across all cells")
        res=adata.X
        #avoid all 0 columns
        mean0,var0=get_mean_var(res)
        sd0=np.sqrt(var0)
        sd0[sd0<=1e-5]=1e-5
        if issparse(res):
            res=res.toarray()
        res-=mean0
        res/=sd0
        res[res>max_value]=max_value
        adata.X=res
    return adata
```

å†è¯•ä¸€ä¸‹

```python
scale_bygroup(adata_all, groupby='batch', max_value=6)
```

å¯èƒ½æŠ¥å¦å¤–ä¸€ä¸ªå…³äºtensorçš„é”™

å†åŠ ä¸€è¡Œ

```python
adata_all.X = adata_all.X.todense()
```

è¿™æ ·åº”è¯¥å°±æ²¡é—®é¢˜äº†ğŸ¤£

### 5.2 desc

```PYTHON
adata_all = desc.train(adata_all, dims=[adata_all.shape[1], 512, 64], tol=0.005, 
										n_neighbors=10,
										batch_size=256, 
										louvain_resolution= [1.0],#å¯ç”¨å¤šä¸ªåˆ†è¾¨ç‡
                    save_dir="seurat", do_tsne=True, learning_rate=300,
                    do_umap=True, num_Cores_tsne=4,
                    save_encoder_weights=True)
```

ä¸‰ä¸ªé‡è¦è¾“å‡ºç»“æœ 

èšç±»ç»“æœï¼šadata_all.obs['desc_1.0']

é«˜è´¨é‡ï¼ˆç›¸è¾ƒPCAæ¥è¯´ï¼‰ä½ç»´è¡¨è¾¾ï¼šadata_all.obsm['X_Embeded_z1.0']

UMAPåæ ‡ï¼ˆç”¨äºå¯è§†åŒ–ï¼‰ï¼šadata_all.obsm['X_umap1.0']

### 5.3 ä¿å­˜ç»“æœ

```python
adata_all.write('desc_result_batch_hvg2000_res1.h5ad')
```

### 5.4 è‡ªå®šä¹‰åˆ†ç¾¤

```python
adata_all.obs['type'] = 'B'

for i in range(1,3):
    name = 'Donor'+str(i)
    adata_all.obs.loc[adata_all.obs_names.str.endswith(name),'type'] = 'A'

adata_all.write('desc_result_batch_hvg2000_res1.h5ad')
```

### 5.5 å¯è§†åŒ–

```python
with rc_context({'figure.figsize': (18, 12)}):
  sc.pl.scatter(adata_all,basis='umap1.0',
                color='desc_1.0',save='_batch_hvg2000_res1_desc.png',
                legend_fontsize = 'x-large', size = 7)
```

## Part VI åšèœâ€”â€”ä¸‹æ¸¸åˆ†æï¼ˆéƒ½æ˜¯å‘å•Šï¼‰

### 6.1 è½¬åŒ–h5adæ•°æ®è‡³seuratå¯¹è±¡

Seurat4éœ€è¦R4.1.0ï¼Œæˆ‘ä»¬å…ˆç”¨R4.1.0

```R
library(Seurat)
library(SeuratDisk)
library(ggplot2)
library(plyr)
```

```R
Convert("desc_result_batch_hvg2000_res1.h5ad", "desc_result_batch_hvg2000_res1.h5seurat")
seuratObject <- LoadH5Seurat("desc_result_batch_hvg2000_res1.h5seurat")
saveRDS(seuratObject, "seuratObject.rds")
```

ä¸Šè¿°èƒ½å¦è·‘é€šçœŸçš„æ˜¯è¦é è¿æ°”äº†ï¼Œæ— æƒ…å•Šï¼

### 6.2 æ ¹æ®éœ€æ±‚æŒ‰è‡ªå®šä¹‰çš„identsèšç±»å¯è§†åŒ–

identså°±æ˜¯è‡ªå®šä¹‰çš„ç»†èƒç¾¤groupï¼Œä¹Ÿå¯ä»¥ç”¨seuratæ“ä½œã€‚

```R
seuratObject <- NormalizeData(seuratObject)#è½¬æ¢æ˜¯ç”¨çš„rawï¼Œéœ€è¦å†æ¬¡æ ‡å‡†åŒ–
group = c('desc_1.0','batch','type','stage')
for (i in group){
    name <- paste(i,'.png')
    p <- DimPlot(seuratObject, group.by = i,reduction = 'umap1.0',label = TRUE,pt.size = 0.05)
    ggsave(plot = p, filename = name,width = 9,height = 6)
}
Idents(object = seuratObject) <- seuratObject@meta.data$`desc_1.0`#è¿™æ ·ä»¥åä¸ç”¨group.byï¼Œé»˜è®¤æ˜¯è¿™ä¸ªidents
```

u1s1ï¼Œggplot2ç”»å›¾å°±æ˜¯é¦™å•Šã€‚

### 6.3 å·®å¼‚è¡¨è¾¾åˆ†æ

- FindAllMarkersï¼Œæ¯ä¸ªç¾¤æŒ‰avg_log2FCå€¼å–å‰10ä¸ªåŸºå› 

```R
seuratObject.markers <- FindAllMarkers(seuratObject, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
write.table(x=seuratObject.markers,file = '',quote = 'false', sep = '\t')
top10 <- seuratObject.markers %>% group_by(cluster) %>% top_n(10, avg_log2FC)
write.csv(top10,'top.csv')
```

- FindMarkers

```R
for (i in c(0:1)){
    f <- FindMarkers(object = seuratObject,`ident.1` = 'A',group.by = 'type',subset.ident = i)
    de <- f %>% filter(avg_log2FC>0 & p_val_adj<0.05 )#è¿™é‡Œä¸åŒ
    g <- as.data.frame(rownames(de))
    name <- paste('up',as.character(i),'.csv',sep = '')
    write.table(g,name,sep = ',',quote = FALSE,row.names = FALSE,col.names = as.character(i))
}
for (i in c(3:4)){
    f <- FindMarkers(object = seuratObject,`ident.1` = 'A',group.by = 'type',subset.ident = i)
    de <- f %>% filter(avg_log2FC<0 & p_val_adj<0.05 )#è¿™é‡Œä¸åŒ
    g <- as.data.frame(rownames(de))
    name <- paste('down',as.character(i),'.csv',sep = '')
    write.table(g,name,sep = ',',quote = FALSE,row.names = FALSE,col.names = as.character(i))
}
```

### 6.4 å¯Œé›†åˆ†æ

æŠŠåˆšæ‰çš„csvæ–‡ä»¶æ”¾åœ¨ä¸‹é¢ç½‘ç«™é‡Œé¢æ‰‹åŠ¨ç‚¹ç‚¹ç‚¹

https://metascape.org/gp/index.html#/main/step1

ç„¶åæŠŠç»“æœç»™æ‡‚ç”Ÿç‰©å­¦çš„äººï¼ˆè¿™æˆ‘å®Œå…¨ä¸æ‡‚ğŸŒ¶ğŸ¤ªï¼‰

### 6.5 ç»†èƒç¾¤æ ‡æ³¨

ä¹Ÿå°±æ˜¯ç»™åˆšåˆšçš„æ— ç›‘ç£èšç±»ç»“æœè´´ä¸Šæ ‡ç­¾ã€‚

æ‡‚ç”Ÿç‰©å­¦çš„äººåˆ†æå¾—åˆ°ç»†èƒç¾¤åç§°ï¼Œç”¨plyr::mapvalueæ–°åŠ ä¸€åˆ—identså³å¯ã€‚

### 6.6 æ‹Ÿæ—¶è½¨è¿¹åˆ†æï¼ˆç¬¬ä¸‰ä¸ªèœœæ±å¤§å‘ï¼‰

æˆ‘ä»¬å°†åˆšåˆšå¾—åˆ°çš„seuratObjectä¿å­˜

```R
saveRDS(seuratObject, "seuratObject.rds")
```

å†æ–°å»ºä¸€ä¸ªR3.3.6çš„ç¯å¢ƒ

```R
library(Seurat)
library(ggplot2)
library(dplyr)
# library(SeuratObject)
library(monocle3)#æˆ‘ç”¨çš„0.2.3
```

```R
cds <- new_cell_data_set(expdata,cell_metadata = cell_metadata,gene_metadata = gene_annotation)

# cds <- preprocess_cds(cds,method = 'PCA',num_dim = 30)
#å¦‚æœç”¨ä»–çš„PCAæŠ¥æœ‰å…³irlbaçš„é”™è¯¯ï¼Œå¯ä»¥é™ä¸€ä¸‹ç‰ˆæœ¬
# cds <- reduce_dimension(cds)

reducedDims(cds)$PCA <- as.data.frame(dat@reductions$Embeded_z1.0@cell.embeddings)
#å¯ä»¥ä¸ç”¨ä»–çš„PCA
reducedDims(cds)$UMAP <- as.data.frame(dat@reductions$umap1.0@cell.embeddings)
#å¯ä»¥ä¸ç”¨ä»–çš„UMAP
cds <- cluster_cells(cds,reduction_method = 'UMAP')

cds <- learn_graph(cds,use_partition = FALSE)#use_partitioné»˜è®¤æ˜¯TRUE,å¦‚æœä¹‹åçš„plot_cells æœ‰ç°è‰²éƒ¨åˆ†å°±è®¾ç½®æˆFALSE

cds <- order_cells(cds,root_cells = rownames(subset(cds@colData,desc_1.0=='2')))
#éœ€è¦æ‰‹åŠ¨è®¾ç½®ç»†èƒèŠ‚ç‚¹
plot_cells(cds,
           color_cells_by = "pseudotime",
           label_cell_groups=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           graph_label_size=1.5)
```

æœ€ç»ˆå›¾ç‰‡

![img](https://cole-trapnell-lab.github.io/monocle3/images/manual_images/embryo_pr_graph_by_pseudotime_programmatically_ordered.png)

å½“ç„¶å¯ä»¥ç”¨è‡ªå·±åˆ†çš„ç»†èƒç¾¤å†ç”»ä¸‹ã€‚

å‘è­¦å‘Šï¼šR4.1.0ä¼šåœ¨learn_graphæŠ¥å„å¼å„æ ·çš„é”™è¯¯ï¼Œå°šæœªæœ‰è§£å†³æ–¹æ¡ˆï¼Œæ— æƒ…å“ˆæ‹‰å°‘ï¼

## VII å°¾å£°

åƒè¨€ä¸‡è¯­æ±‡æˆä¸€å¥æ„Ÿæ©ï¼Œæ„Ÿè°¢ã€‚å¸Œæœ›ä»¥åèƒ½å¤Ÿæœ‰æœºä¼šç»§ç»­åšå•ç»†èƒåˆ†æã€‚æœªå®Œå¾…ç»­~
